proposals/0522-or-patterns.rst

```diff
@@ -0,0 +1,354 @@
+
+        urk pat = pprPanic "isIrrefutableHsPat:" (ppr pat)
+
+  Using or patterns this code can be simplified to: ::
+
+    isIrrefutableHsPat pat
+      = go pat
+      where
+        go (L _ pat) = go1 pat
+
+        go1 (WildPat{} ; VarPat{} ; LazyPat{})
+          = True
+
+        go1 (PArrPat{} ; ConPatIn{} ; LitPat{} ; NPat{} ; NPlusKPat{} ; ListPat {})
+          = False
+
+        go1 (BangPat pat)       = go pat
+        go1 (CoPat _ pat _)     = go1 pat
+        go1 (ParPat pat)        = go pat
+        go1 (AsPat _ pat)       = go pat
+        go1 (ViewPat _ pat _)   = go pat
+        go1 (SigPatIn pat _)    = go pat
+        go1 (SigPatOut pat _)   = go pat
+        go1 (CoPat _ pat _)     = go1 pat
+        go1 (TuplePat pats _ _) = all go pats
+        go1 (ConPatOut{ pat_con = L _ (RealDataCon con), pat_args = details }) = ...
+        go1 (ConPatOut{ pat_con = L _ (PatSynCon _pat) }) = ...
+        go1 (SplicePat {})      = urk pat
+
+        urk pat = pprPanic "isIrrefutableHsPat:" (ppr pat)
+
+
+GHC also has wildcard patterns in many places (here  ``Core.hs``):
+
+::
+
+ hasCoreUnfolding (CoreUnfolding {}) = True
+ hasCoreUnfolding (DFunUnfolding {}) = True
+ hasCoreUnfolding _                  = False
+
+ isValueUnfolding (CoreUnfolding { uf_is_value = is_evald }) = is_evald
+ isValueUnfolding _                                          = False
+
+ isEvaldUnfolding (OtherCon _)                               = True
+ isEvaldUnfolding (CoreUnfolding { uf_is_value = is_evald }) = is_evald
+ isEvaldUnfolding _                                          = False
+
+ isConLikeUnfolding (OtherCon _)                             = True
+ isConLikeUnfolding (CoreUnfolding { uf_is_conlike = con })  = con
+ isConLikeUnfolding _                                        = False
+
+ hasSomeUnfolding NoUnfolding   = False
+ hasSomeUnfolding BootUnfolding = False
+ hasSomeUnfolding _             = True
+
+ neverUnfoldGuidance UnfNever = True
+ neverUnfoldGuidance _        = False                                           
+
+ ...
+
+Would ``Unfolding`` be expanded by another constructor, all these functions would still compile but some would become semantically wrong, laying an additional burden on the code author.
+
+
+Effect and Interactions
+-----------------------
+
+The main effect of or patterns is twofold:
+
+1. With or patterns developers can avoid ``_`` wildcard patterns which can
+   unintentionally match constructors as types are being extended.
+
+2. Or patterns allow more code reuse as right hand sides can be shared by many patterns.
+
+
+GADTs & Existential quantification
+~~~~~~~~~~~~~~~~~
+
+With existential quantification and GADTs, patterns can not only bind values, but also equality constraints, dictionaries and existential type variables. This however does not interfere with or patterns in any way - binding variables (or equality constraints etc.) explicitly is a non-goal of this proposal.
```

Comment:
I think you still need to clarify this, even if the constranits are not named, the question whether they are available still applies.

So how about this:
```
data Foo a where
  Foo1 : Eq a => Foo a
  Foo2 : Eq a => Foo a
```
will this work?
```
foo :: a -> Foo a -> Bool
foo x (Foo1{} ; Foo2 {}) = x == x
```

Reply:
Not binding variables is consistent with a hypothetical future proposal that *does* bind variables, since that would simply allow more programs.

However, binding or not binding constraints results in two incompatible solutions, I believe. I think that might be worth explicitly pointing out.

Reply:
I think “not binding constraints” is compatible with a future that binds the (intersection of) the constraints.  You can only write less programs with less constraints available. 

(Ok, some programs might in the bright future give warnings about dead code when the constraints are actually unsatisfiable; we can probably live with that.)

You should also clarify what happens here:
```
data GADT a where IsInt1 :: GADT Int; IsInt2 :: GADT Int

foo :: a -> GADT -> a
foo x (IsInt1 {}; IsInt 2 {}) = x + 1
```

This is very much similar to the example above when one thinks of GADTs as constructors binding equality constraints, but maybe that thinking is not universal, so it’d be good to clarify both separately.

Reply:
@nomeata ah, you're probably right – my thought process was essentially that with variables, the typical idea is that all patterns have to bind the same variables of the same type, so I was thinking all patterns also have to bind the same constraints. But as you say, you could bind the intersection instead.